### PR TITLE
Implement quantum state reconstruction (pure & mixed)

### PR DESCRIPTION
**Pull request by JasonZHM**
This pull request implements the quantum state reconstruction driver which minimizes the KL divergence to approximate the target state. 

Features:
- compatible with both pure and mixed ansatz
- implement a control variates method (stochastic variance reduced gradient, SVRG) to reduce the variance caused by mini-batch training
- support chunking when computing control variate expectation
- an example of reconstructing 1d Ising chain ground state
- tests on pure & mixed state reconstruction and KL computation

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Very nice! the rendered documentation is here https://netket--1427.org.readthedocs.build/en/1427/api/_generated/driver/netket.driver.QSR.html#netket.driver.QSR

**Comment by JasonZHM** [comment]
> Very nice! the rendered documentation is here https://netket--1427.org.readthedocs.build/en/1427/api/_generated/driver/netket.driver.QSR.html#netket.driver.QSR

Looks good! I just fixed the url and a few typos in the doc. Anything else to do?

File: Examples/StateReconstruction/ising_1d_qsr.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import netket as nk
+import numpy as np
+from netket.driver import QSR
```
Comment (femtobit) [inline Examples/StateReconstruction/ising_1d_qsr.py:3]: Hi @JasonZHM, thanks for the contribution! One minor thing: Typically, main NetKet drivers are exposed directly under the `netket` top-level module for convenience (i.e., as `netket.QSR` in this case).

There is no need to move the existing code, you can just add an import here (and update the example):
https://github.com/netket/netket/blob/a916167944ba6df7394b6a7274fbfdb5c6550a4b/netket/__init__.py#L61-L63

File: netket/driver/qsr.py [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+        if len(target_state.shape) == 1:
+            target_state = target_state.reshape(-1, 1)
+            target_state = target_state @ target_state.conj().T
+        rotations = self._training_rotations[::n_shots]
+        KL_list = []
+        try:
+            vs = self.state.to_matrix(normalize=True)
+        except:
+            vs = self.state.to_array(normalize=True).reshape(-1, 1)
+            vs = vs @ vs.conj().T
```
Comment (femtobit) [inline netket/driver/qsr.py:984]: Flake complains about the `except` here, which does not specify a specific exception type to catch. In any case, I think this logic might be better realized without replying on exception handling. A simple alternative could be to check the type of `self.state` via `isinstance` in order to only call `to_matrix` on a type that supports it.

**Review by femtobit — COMMENTED** [review]
There are a couple minor issues raised by our [automated code check](https://github.com/netket/netket/actions/runs/4126857251/jobs/7129324700) using flake, mostly formatting errors (did you run [`black`](https://github.com/psf/black) on the code?) and some unused imports/variables. Cleaning those up should make the code checker happy.

```
Examples/StateReconstruction/basis_generators.py:5:1: E302 expected 2 blank lines, found 1
Examples/StateReconstruction/generate_data.py:13:5: F841 local variable 'N' is assigned to but never used
Examples/StateReconstruction/ising_1d_qsr.py:56:1: E302 expected 2 blank lines, found 1
netket/driver/qsr.py:31:1: F401 'netket.nn' imported but unused
netket/driver/qsr.py:37:1: F401 'netket.utils.types.PyTree' imported but unused
netket/driver/qsr.py:37:1: F401 'netket.utils.types.Scalar' imported but unused
netket/driver/qsr.py:324:1: F811 redefinition of unused '_grad_negative' from line 289
netket/driver/qsr.py:486:80: W291 trailing whitespace
netket/driver/qsr.py:502:154: W291 trailing whitespace
netket/driver/qsr.py:507:70: W291 trailing whitespace
netket/driver/qsr.py:508:1: W293 blank line contains whitespace
netket/driver/qsr.py:509:14: W291 trailing whitespace
netket/driver/qsr.py:511:41: W291 trailing whitespace
netket/driver/qsr.py:515:1: W293 blank line contains whitespace
netket/driver/qsr.py:982:9: E722 do not use bare 'except'
```

File: netket/driver/qsr.py (hunk 2 above)
Comment (JasonZHM) [reply netket/driver/qsr.py]: Thanks for pointing these out. I did use black, but it seems that these problems were not reported. I've fixed the formatting and removed the try-except.

**Comment by PhilipVinc** [comment]
@JasonZHM that looks great, thank you!

Just one last thing. 
As the file is huge...
 could you please move this driver to a new module in the experimental submodule, `netket.experimental.qsr` and split the driver and supporting logic in a few files?

Then you can import the `QSR` driver in `netket.experimental.__init__`.
If you then fix the Examples as well as the documentation, it's good to go.

Thanks a lot!

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
I refactored a bit the code to also simplify the internal logic so that it's more compact and there is a better separation of the various pieces, creating an (internal) `dataset` object.

This should make the code slightly more readable as well...

Thank you @JasonZHM for this contribution! It was a long time since we wanted to implement this again...

**Pull request merged**